File: ai/calendar/sriLankaCalendar.js
/**
 * Sri Lankan calendar features for demand forecasting inference.
 * Must produce the same feature vector as ai/calendar/sriLankaCalendar.py
 * so training and inference stay aligned.
 */
const fs = require("fs");
const path = require("path");

/** @type {any} */
const events = JSON.parse(
  fs.readFileSync(path.join(__dirname, "events.json"), "utf8"),
);

const FEATURE_NAMES = Object.freeze([
  "dayOfWeek",
  "dayOfMonth",
  "month",
  "isPublicHoliday",
  "isPoyaDay",
  "isFestivalDay",
  "daysToNextFestival",
  "daysSinceLastFestival",
  "inSinhalaTamilNewYearWindow",
  "inEidAlFitrWindow",
  "inDeepavaliWindow",
  "inChristmasWindow",
  "inSalaryWindow",
  "inMonsoonYala",
  "inMonsoonMaha",
]);

/** @param {Date} d */
function toIsoDate(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}-${m}-${day}`;
}

/** @param {string} s */
function parseIso(s) {
  const [y, m, d] = s.split("-").map(Number);
  return new Date(y, m - 1, d);
}

/**
 * @param {Date} a
 * @param {Date} b
 * @returns {number} whole days from a to b (b - a)
 */
function daysBetween(a, b) {
  return Math.round((b.getTime() - a.getTime()) / 86_400_000);
}

/** @param {Date} d */
function isPoyaDay(d) {
  const list = events.poyaDays?.[String(d.getFullYear())] || [];
  return Array.isArray(list) && list.includes(toIsoDate(d));
}

/** @param {Date} d */
function isFixedHoliday(d) {
  const month = d.getMonth() + 1;
  const day = d.getDate();
  return (events.fixedPublicHolidays || []).some(
    (/** @type {{month:number, day:number}} */ h) =>
      h.month === month && h.day === day,
  );
}

/** @param {Date} d */
function isSinhalaTamilNewYear(d) {
  const cfg = events.sinhalaTamilNewYear;
  return d.getMonth() + 1 === cfg.month && cfg.days.includes(d.getDate());
}

/** @param {number} year */
function getEidAlFitr(year) {
  const entry = events.islamicFestivals?.eidAlFitr?.[String(year)];
  return entry?.date ? parseIso(entry.date) : null;
}

/** @param {number} year */
function getEidAlFitrPreWindow(year) {
  const entry = events.islamicFestivals?.eidAlFitr?.[String(year)];
  return entry?.preWindowDays ?? 14;
}

/** @param {number} year */
function getEidAlAdha(year) {
  const iso = events.islamicFestivals?.eidAlAdha?.[String(year)];
  return typeof iso === "string" ? parseIso(iso) : null;
}

/** @param {number} year */
function getDeepavali(year) {
  const iso = events.deepavali?.[String(year)];
  return typeof iso === "string" ? parseIso(iso) : null;
}

/** @param {number} year */
function getChristmas(year) {
  return new Date(year, 11, 25);
}

/** @param {number} year */
function getSinhalaTamilNYStart(year) {
  const cfg = events.sinhalaTamilNewYear;
  return new Date(year, cfg.month - 1, cfg.days[0]);
}

/** @param {number} year */
function majorFestivalDates(year) {
  const dates = [getSinhalaTamilNYStart(year), getChristmas(year)];
  const eidF = getEidAlFitr(year);
  if (eidF) dates.push(eidF);
  const eidA = getEidAlAdha(year);
  if (eidA) dates.push(eidA);
  const dpv = getDeepavali(year);
  if (dpv) dates.push(dpv);
  return dates;
}

/**
 * @param {Date} d
 * @param {Date | null} festival
 * @param {number} preDays
 * @param {number} postDays
 */
function inWindow(d, festival, preDays, postDays) {
  if (!festival) return false;
  const diff = daysBetween(festival, d);
  return diff >= -preDays && diff <= postDays;
}

/** @param {Date} d */
function isPublicHoliday(d) {
  const year = d.getFullYear();
  if (isFixedHoliday(d)) return true;
  if (isSinhalaTamilNewYear(d)) return true;
  if (isPoyaDay(d)) return true;
  const eidF = getEidAlFitr(year);
  if (eidF && daysBetween(eidF, d) === 0) return true;
  const eidA = getEidAlAdha(year);
  if (eidA && daysBetween(eidA, d) === 0) return true;
  const dpv = getDeepavali(year);
  if (dpv && daysBetween(dpv, d) === 0) return true;
  return false;
}

/** @param {Date} d */
function inMonsoonYala(d) {
  const m = d.getMonth() + 1;
  const cfg = events.monsoonSeasons.yala;
  return m >= cfg.startMonth && m <= cfg.endMonth;
}

/** @param {Date} d */
function inMonsoonMaha(d) {
  const m = d.getMonth() + 1;
  const cfg = events.monsoonSeasons.maha;
  return m >= cfg.startMonth || m <= cfg.endMonth;
}

/** @param {Date} d */
function inSalaryWindow(d) {
  const day = d.getDate();
  const cfg = events.salaryWindow;
  return day >= cfg.startDay || day <= cfg.endDay;
}

/** @param {Date} d */
function daysToNearestFestival(d) {
  const year = d.getFullYear();
  const all = [
    ...majorFestivalDates(year - 1),
    ...majorFestivalDates(year),
    ...majorFestivalDates(year + 1),
  ];
  let next = 999;
  let prev = 999;
  for (const f of all) {
    const diff = daysBetween(d, f);
    if (diff >= 0 && diff < next) next = diff;
    if (diff < 0 && -diff < prev) prev = -diff;
  }
  return { daysToNextFestival: next, daysSinceLastFestival: prev };
}

/**
 * Build the calendar feature dictionary for a given date.
 * @param {Date | string} date
 */
function getCalendarFeatures(date) {
  const d = date instanceof Date ? new Date(date.getTime()) : parseIso(date);
  d.setHours(0, 0, 0, 0);

  const year = d.getFullYear();
  const stny = getSinhalaTamilNYStart(year);
  const eidF = getEidAlFitr(year);
  const dpv = getDeepavali(year);
  const xmas = getChristmas(year);

  const isFestivalDay =
    isSinhalaTamilNewYear(d) ||
    (eidF && daysBetween(eidF, d) === 0) ||
    (dpv && daysBetween(dpv, d) === 0) ||
    daysBetween(xmas, d) === 0;

  const { daysToNextFestival, daysSinceLastFestival } =
    daysToNearestFestival(d);

  return {
    dayOfWeek: d.getDay(),
    dayOfMonth: d.getDate(),
    month: d.getMonth() + 1,
    isPublicHoliday: isPublicHoliday(d) ? 1 : 0,
    isPoyaDay: isPoyaDay(d) ? 1 : 0,
    isFestivalDay: isFestivalDay ? 1 : 0,
    daysToNextFestival,
    daysSinceLastFestival,
    inSinhalaTamilNewYearWindow: inWindow(
      d,
      stny,
      events.sinhalaTamilNewYear.preWindowDays,
      2,
    )
      ? 1
      : 0,
    inEidAlFitrWindow: inWindow(d, eidF, getEidAlFitrPreWindow(year), 2)
      ? 1
      : 0,
    inDeepavaliWindow: inWindow(d, dpv, 14, 2) ? 1 : 0,
    inChristmasWindow: inWindow(d, xmas, events.christmas.preWindowDays, 1)
      ? 1
      : 0,
    inSalaryWindow: inSalaryWindow(d) ? 1 : 0,
    inMonsoonYala: inMonsoonYala(d) ? 1 : 0,
    inMonsoonMaha: inMonsoonMaha(d) ? 1 : 0,
  };
}

/**
 * Build the calendar feature vector in FEATURE_NAMES order.
 * @param {Date | string} date
 */
function getFeatureVector(date) {
  const features = getCalendarFeatures(date);
  return FEATURE_NAMES.map(
    (name) => /** @type {any} */ (features)[name],
  );
}

module.exports = {
  getCalendarFeatures,
  getFeatureVector,
  getFeatureNames: () => [...FEATURE_NAMES],
};

File: ai/calendar/sriLankaCalendar.py
"""Sri Lankan calendar features for demand forecasting training.

Mirrors ai/calendar/sriLankaCalendar.js — feature names, order, and values
must match exactly so a model trained here can be served by the JS inference
runtime.
"""
import json
from datetime import date, datetime
from pathlib import Path
from typing import Optional, Union

_EVENTS_PATH = Path(__file__).parent / "events.json"
with _EVENTS_PATH.open(encoding="utf-8") as _f:
    EVENTS = json.load(_f)

FEATURE_NAMES = (
    "dayOfWeek",
    "dayOfMonth",
    "month",
    "isPublicHoliday",
    "isPoyaDay",
    "isFestivalDay",
    "daysToNextFestival",
    "daysSinceLastFestival",
    "inSinhalaTamilNewYearWindow",
    "inEidAlFitrWindow",
    "inDeepavaliWindow",
    "inChristmasWindow",
    "inSalaryWindow",
    "inMonsoonYala",
    "inMonsoonMaha",
)


def _to_date(d: Union[str, date, datetime]) -> date:
    if isinstance(d, datetime):
        return d.date()
    if isinstance(d, date):
        return d
    return datetime.strptime(d, "%Y-%m-%d").date()


def _parse_iso(s: str) -> date:
    return datetime.strptime(s, "%Y-%m-%d").date()


def _is_poya_day(d: date) -> bool:
    return d.isoformat() in EVENTS.get("poyaDays", {}).get(str(d.year), [])


def _is_fixed_holiday(d: date) -> bool:
    return any(
        h["month"] == d.month and h["day"] == d.day
        for h in EVENTS.get("fixedPublicHolidays", [])
    )


def _is_sinhala_tamil_ny(d: date) -> bool:
    cfg = EVENTS["sinhalaTamilNewYear"]
    return d.month == cfg["month"] and d.day in cfg["days"]


def _get_eid_al_fitr(year: int) -> Optional[date]:
    entry = EVENTS.get("islamicFestivals", {}).get("eidAlFitr", {}).get(str(year))
    if isinstance(entry, dict) and "date" in entry:
        return _parse_iso(entry["date"])
    return None


def _get_eid_al_fitr_pre_window(year: int) -> int:
    entry = EVENTS.get("islamicFestivals", {}).get("eidAlFitr", {}).get(str(year))
    if isinstance(entry, dict):
        return int(entry.get("preWindowDays", 14))
    return 14


def _get_eid_al_adha(year: int) -> Optional[date]:
    iso = EVENTS.get("islamicFestivals", {}).get("eidAlAdha", {}).get(str(year))
    return _parse_iso(iso) if isinstance(iso, str) else None


def _get_deepavali(year: int) -> Optional[date]:
    iso = EVENTS.get("deepavali", {}).get(str(year))
    return _parse_iso(iso) if isinstance(iso, str) else None


def _get_christmas(year: int) -> date:
    return date(year, 12, 25)


def _get_sinhala_tamil_ny_start(year: int) -> date:
    cfg = EVENTS["sinhalaTamilNewYear"]
    return date(year, cfg["month"], cfg["days"][0])


def _major_festival_dates(year: int) -> list:
    dates = [_get_sinhala_tamil_ny_start(year), _get_christmas(year)]
    for getter in (_get_eid_al_fitr, _get_eid_al_adha, _get_deepavali):
        v = getter(year)
        if v is not None:
            dates.append(v)
    return dates


def _in_window(
    d: date, festival: Optional[date], pre_days: int, post_days: int
) -> bool:
    if festival is None:
        return False
    diff = (d - festival).days
    return -pre_days <= diff <= post_days


def _is_public_holiday(d: date) -> bool:
    if _is_fixed_holiday(d) or _is_sinhala_tamil_ny(d) or _is_poya_day(d):
        return True
    year = d.year
    if d == _get_eid_al_fitr(year):
        return True
    if d == _get_eid_al_adha(year):
        return True
    if d == _get_deepavali(year):
        return True
    return False


def _in_monsoon_yala(d: date) -> bool:
    cfg = EVENTS["monsoonSeasons"]["yala"]
    return cfg["startMonth"] <= d.month <= cfg["endMonth"]


def _in_monsoon_maha(d: date) -> bool:
    cfg = EVENTS["monsoonSeasons"]["maha"]
    return d.month >= cfg["startMonth"] or d.month <= cfg["endMonth"]


def _in_salary_window(d: date) -> bool:
    cfg = EVENTS["salaryWindow"]
    return d.day >= cfg["startDay"] or d.day <= cfg["endDay"]


def _days_to_nearest_festival(d: date) -> tuple:
    all_dates = []
    for y in (d.year - 1, d.year, d.year + 1):
        all_dates.extend(_major_festival_dates(y))
    next_d, prev_d = 999, 999
    for f in all_dates:
        diff = (f - d).days
        if diff >= 0:
            next_d = min(next_d, diff)
        else:
            prev_d = min(prev_d, -diff)
    return next_d, prev_d


def get_calendar_features(d: Union[str, date, datetime]) -> dict:
    """Return the calendar feature dictionary for a given date."""
    d = _to_date(d)
    year = d.year
    stny = _get_sinhala_tamil_ny_start(year)
    eid_fitr = _get_eid_al_fitr(year)
    deepavali = _get_deepavali(year)
    xmas = _get_christmas(year)

    is_festival_day = (
        _is_sinhala_tamil_ny(d)
        or d == eid_fitr
        or d == deepavali
        or d == xmas
    )

    days_to_next, days_since_last = _days_to_nearest_festival(d)

    # JS Date.getDay() returns Sun=0..Sat=6. Python weekday() is Mon=0..Sun=6.
    day_of_week_js = (d.weekday() + 1) % 7

    return {
        "dayOfWeek": day_of_week_js,
        "dayOfMonth": d.day,
        "month": d.month,
        "isPublicHoliday": int(_is_public_holiday(d)),
        "isPoyaDay": int(_is_poya_day(d)),
        "isFestivalDay": int(is_festival_day),
        "daysToNextFestival": days_to_next,
        "daysSinceLastFestival": days_since_last,
        "inSinhalaTamilNewYearWindow": int(
            _in_window(d, stny, EVENTS["sinhalaTamilNewYear"]["preWindowDays"], 2)
        ),
        "inEidAlFitrWindow": int(
            _in_window(d, eid_fitr, _get_eid_al_fitr_pre_window(year), 2)
        ),
        "inDeepavaliWindow": int(_in_window(d, deepavali, 14, 2)),
        "inChristmasWindow": int(
            _in_window(d, xmas, EVENTS["christmas"]["preWindowDays"], 1)
        ),
        "inSalaryWindow": int(_in_salary_window(d)),
        "inMonsoonYala": int(_in_monsoon_yala(d)),
        "inMonsoonMaha": int(_in_monsoon_maha(d)),
    }


def get_feature_vector(d: Union[str, date, datetime]) -> list:
    """Return the feature vector in FEATURE_NAMES order."""
    features = get_calendar_features(d)
    return [features[name] for name in FEATURE_NAMES]


def get_feature_names() -> list:
    return list(FEATURE_NAMES)
